File: DDDSample.Domain/Aggregates/Customer/CustomerState.vb
Imports DDDSample

Public Class CustomerState
    Private _id As CustomerId
    Private _name As String
    Private _created As Boolean

    Public Sub New(events As IEnumerable(Of IEvent))
        For Each [event] In events
            Mutate([event])
        Next
    End Sub

    Public ReadOnly Property Created As Boolean
        Get
            Return _created
        End Get
    End Property
    Public Sub Mutate([event] As IEvent)
        Me.When([event])
    End Sub
    Public Sub [When]([event] As CustomerCreated)
        Me.ID = [event].Id
        Me.Name = [event].CustomerName
        Me._created = True
    End Sub

    Public Property ID As CustomerId
        Get
            Return _id
        End Get
        Protected Set(value As CustomerId)
            _id = value
        End Set
    End Property
    Public Property Name As String
        Get
            Return _name
        End Get
        Protected Set(value As String)
            _name = value
        End Set
    End Property
End Class

File: DDDSample.Domain/Aggregates/Customer/CustomerState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DDDSample
{
    public class CustomerState
    {
        private CustomerId _id; 
        private string _name;
        private bool _created;

        public CustomerState(IEnumerable<IEvent> events)
        {
            foreach (var e in events)
            {
                Mutate(e);
            }
        }
        public bool Created {get {return _created;}} 
        

        public void Mutate(IEvent e)
        {
            ((dynamic) this).When((dynamic)e);
        }
        
        public  void When(CustomerCreated e)
        {
            this.Id = e.Id;
            this.Name = e.CustomerName;
            this._created = true;
        }
       
        public CustomerId Id{get;private set;}
        public string Name {get;private set;}
    }
}
